build-tools/packages/build-cli/src/commands/generate/packlist.ts

```diff
@@ -0,0 +1,68 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "node:path";
+import type { Package } from "@fluidframework/build-tools";
+import Arborist from "@npmcli/arborist";
+import { Flags } from "@oclif/core";
+import { writeFile } from "fs-extra";
+import packlist from "npm-packlist";
+import { PackageCommand } from "../../BasePackageCommand";
+
+/**
+ * Outputs a list of files that will be included in a package based on its 'files' property in package.json and any
+ * npmignore files.
+ */
+export default class GeneratePackListCommand extends PackageCommand<
+	typeof GeneratePackListCommand
+> {
+	static readonly description =
+		"Outputs a list of files that will be included in a package based on its 'files' property in package.json and any .npmignore files.";
+
+	static readonly flags = {
+		out: Flags.file({
+			description:
+				"File to output the pack list to. This path is relative to the package whose contents is being listed.",
+			default: "packlist.txt",
+			exists: false,
+		}),
+		...PackageCommand.flags,
+	} as const;
+
+	protected async processPackage(pkg: Package): Promise<void> {
+		const { out } = this.flags;
+
+		const outFile = path.join(pkg.directory, out);
+		const arborist = new Arborist({ path: pkg.directory });
+		const tree = await arborist.loadActual();
+		const files = await packlist(tree);
```

Comment:
I see the docs for npm-packlist say this is how you invoke it, and it leaves me confused because `arborist.loadActual()` says it "reads the actual contents of node_modules" (which matches its main docs, "inspect and manage node_modules trees"), which shouldn't be relevant to what the packed package contains. So sounds like that's not exactly what it does, and that already leaves a bit of a sour taste in my mouth 😄 .

Would it work for our purposes to just generate the packed packages and use tar to list their contents?

```
[6][08:58:18] tar --gzip --list --file fluidframework-container-runtime-2.0.0-rc.4.0.0.tgz
package/LICENSE
package/.eslintrc.cjs
package/.mocharc.cjs
package/prettier.config.cjs
package/dist/opLifecycle/batchManager.js
package/lib/opLifecycle/batchManager.js
package/dist/batchTracker.js
package/lib/batchTracker.js
package/dist/blobManager.js
package/lib/blobManager.js
package/dist/channelCollection.js
package/lib/channelCollection.js
package/dist/connectionTelemetry.js
package/lib/connectionTelemetry.js
...
``` 

We'd want to strip the leading `package/` and sort, but I like the idea of "this is 100% guaranteed the actual content, not what some npm package thinks the content will be", so we're not subject to issues if there's a bug in arborist or npm-packlist.

Reply:
`npm pack --dry-run --json` also has what we need, but it seemed less elegant to use exec. But maybe that's better anyway since npm doesn't have an official public api. I agree using arborist here feels strange, even if it does "work."

I'll play around with the other approach.

Reply:
Tar is also not cross platform.

Reply:
I like this version better, basically for the same reason stated above.